### ISSUE TITLE
[Bug]: Warning: React version specified in eslint-plugin-react-settings must be a valid semver version, or "detect"; got “^17”

### ISSUE THREAD

**Issue by Redirts** [jsx-eslint/eslint-plugin-react#3857]
### Is there an existing issue for this?

- [x] I have searched the existing issues and my issue is unique
- [x] My issue appears in the command-line and not only in the text editor

### Description Overview

I am getting this Warning message:

Warning: React version specified in eslint-plugin-react-settings must be a valid semver version, or "detect"; got “^17”

I have a .eslintrc file with this setup:

  "settings": {
    "react": {
      "version": "detect"
    },

### Expected Behavior

This plugin should be able to cope with carets which are part of the semver specification

### eslint-plugin-react version

v7.37.2

### eslint version

v8.57.0

### node version

v20.10.0

**Comment by ljharb** [comment]
Technically no, ranges are not part of the semver specification, they're something npm invented (although, they should be, and there's an open PR on the spec to add them).

It's meant to target a specific version. is there a reason you can't use "detect"?

**Comment by Redirts** [comment]
I currently have "detect" on the .eslintc file. Are you referring to adding that tag somewhere else?

**Comment by ljharb** [comment]
No, that’s what i meant. If you have that setting, where does the ^17 come from?

**Comment by Redirts** [comment]
Well, from what I gather, the "detect" tag in .eslintc is making the plugin fetch the ^17 from my package.json (for the react dependency) and hence giving the warning.

**Comment by ljharb** [comment]
The way "detect" works is it pulls the version number directly from the react you have installed on disk, so package.json shouldn't be involved at all (except in that it informs what ends up on disk).

**Comment by Redirts** [comment]
If not from my package.json then I don't where else it would retrieve ^17 from:
If I do "npm list" I get, for React:

├── react@17.0.2

**Comment by ljharb** [comment]
Can you use `eslint --print-config=path/to/some/file` and see what it says under `settings`? it's possible a shared config or a nested `.eslintrc` file is setting it to `^17`.

**Comment by Redirts** [comment]
![Image](https://github.com/user-attachments/assets/f5d05bd5-3bf0-4785-a2c8-a79530f53e4f)

**Comment by Redirts** [comment]
Is there any news on this @ljharb ? Thanks

**Comment by ljharb** [comment]
I still can't reproduce or figure this out. Are you sure the output you gave was from that eslint command and *not* directly from your config in vscode?

**Comment by Redirts** [comment]
![Image](https://github.com/user-attachments/assets/f826454c-87da-4dad-9a02-6d7b8fcdd2ff)

**Comment by ljharb** [comment]
ok, so then the computed config is still not `^17`, and we still don't know where that's coming from. can you run the same command, but pass the path to a file that normally generates this error?

**Comment by Redirts** [comment]
You raised a good point. Locally, if I run eslint ./src/**/*.js, I don't get that warning/error. I haven't figured out the issue with Sonarqube.

Another thing I tried was, changing the .eslintrc to "detect" and changing the react in package.json to ^17 and then trying to change react version in package.json to "^17.0.0".

So for the first scenario, react version in package.json to ^17 and "detect" in .eslinrc, I still get the warning/error from Sonarqube: 

![Image](https://github.com/user-attachments/assets/b1da30c2-9f4d-4ee8-a220-6e14578f9c60)

If I change the react version in package.json to ^17.0.0 (also "detect" in .eslinrc) I get no complaints from Sonarqube....so it's reading package.json for some reason

**Comment by ljharb** [comment]
Are you using yarn, or pnpm, or something else? Or are you using npm, the standard?

**Comment by Redirts** [comment]
We are only using npm

**Comment by ljharb** [comment]
Please post a link or followup to what you find out from Sonar :-)

**Comment by Redirts** [comment]
Hi @ljharb,

I got this response over on the sonar community forum:

https://community.sonarsource.com/t/warning-react-version-specified-in-eslint-plugin-react-settings-must-be-a-valid-semver-version/133392/7

Can you read it and see if you agree? If not, let me know and I can reply or feel free to reply directly if you have an account there.

Thanks

**Comment by ljharb** [comment]
I don’t; the Sonar person is missing that your config doesn’t contain ^17, your package.json does, and the plugin is reading the version field of react’s package.json on disk, which can’t ever be anything but a full semver version, no ranges.

**Comment by Redirts** [comment]
Thanks for the quick feedback. I have asked him to comment back on this thread to see if we can reach a consensus.

**Comment by Redirts** [comment]
Recognized as a bug on SonarJS: https://sonarsource.atlassian.net/browse/JS-536

**Comment by ljharb** [comment]
Note, that even if ranges become part of the semver specification (there's an open PR to add them), they would never be permitted in the version field of a package.json inside node_modules, so this bug would still exist.
